### PR TITLE
feat: generic JSON based pagination, filter, sort for data tables

### PR DESCRIPTION
**Pull request by sukanya-rath**
## Description:
Adding an admin get reports endpoint with filtering, sorting and pagination

https://finrms.atlassian.net/browse/GEO-665
---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://pay-transparency-pr-522-frontend.apps.silver.devops.gov.bc.ca)
- [Admin Frontend](https://pay-transparency-pr-522-admin-frontend.apps.silver.devops.gov.bc.ca)
- [Backend external API console](https://pay-transparency-pr-522-backend-external.apps.silver.devops.gov.bc.ca/api/V1/docs)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/fin-pay-transparency/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/fin-pay-transparency/actions/workflows/merge.yml)

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://pay-transparency-pr-522-frontend.apps.silver.devops.gov.bc.ca)
- [Admin Frontend](https://pay-transparency-pr-522-admin-frontend.apps.silver.devops.gov.bc.ca)
- [Backend external API console](https://pay-transparency-pr-522-backend-external.apps.silver.devops.gov.bc.ca/api/V1/docs)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/fin-pay-transparency/actions/workflows/analysis.yml)

After merge, new images are deployed in:
- [Merge Workflow](https://github.com/bcgov/fin-pay-transparency/actions/workflows/merge.yml)

### CODE REVIEW

File: backend/src/v1/services/report-search-service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+  ...jest.requireActual('../prisma/prisma-client-readonly-replica'),
+  default: {
+    pay_transparency_report: {
+      findMany: (args) => prismaClient.pay_transparency_report.findMany(args),
+      count: (args) => prismaClient.pay_transparency_report.count(args),
+    },
+  },
+}));
+
+describe('report-search-service', () => {
```
Comment (jer3k) [inline backend/src/v1/services/report-search-service.spec.ts:58]: Can there be some tests for invalid parameters like a key that we don't search on, an invalid object property name, an invalid operation, maybe a value that is expecting an int but instead got a string? Eg. [{key:"user_id", comparator:"bigger", value:"17"}]

File: backend/src/types/report_search.ts
[the remark this answers is not in the document]
Comment (sukanya-rath) [reply backend/src/types/report_search.ts:1]: we are going to do v1

File: backend/src/v1/services/report-search-service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+        expect(response.total).toBe(1);
+      });
+
+      it('neq', async () => {
+        const response = await reportSearchService.searchReport(
+          0,
+          10,
+          '[]',
+          '[{"key": "reporting_year", "operation": "neq", "value": 2024 }]',
+        );
```
Comment (banders) [inline backend/src/v1/services/report-search-service.spec.ts:102]: No "expect" in this test?

File: backend/src/v1/services/report-search-service.spec.ts (hunk 1 above)
Comment (jer3k) [reply backend/src/v1/services/report-search-service.spec.ts:58]: @goemen I just noticed this was done on line 64, sorry didn't notice that earlier.

**Review by banders — APPROVED** [review]

**Review by jer3k — APPROVED** [review]

**Pull request merged**